File: 2-Semestre/AlgoritmosII/APS/aps.c
#include <stdio.h>
//  A biblioteca abaixo torna possível trabalhar com strings em C
#include <string.h>

//  Variáveis globais
char CESTA_BASICA[10][100], SUPERMERCADOS[5][100];
double PRECOS[5][10];

//  Esta função serve para inicializar as variáveis globais
int inicializar(void) {

    //  Atribuição das variáveis globais
    //  CESTA_BASICA
    strcpy(CESTA_BASICA[0], "Arroz");
    strcpy(CESTA_BASICA[1], "Feijao");
    strcpy(CESTA_BASICA[2], "Acucar");
    strcpy(CESTA_BASICA[3], "Oleo");
    strcpy(CESTA_BASICA[4], "Cafe");
    strcpy(CESTA_BASICA[5], "Fuba");
    strcpy(CESTA_BASICA[6], "Ovos");
    strcpy(CESTA_BASICA[7], "Macarrao");
    strcpy(CESTA_BASICA[8], "Leite");
    strcpy(CESTA_BASICA[9], "Frango");

    //  SUPERMERCADOS
    strcpy(SUPERMERCADOS[0], "Carrefour");
    strcpy(SUPERMERCADOS[1], "Guanabara");
    strcpy(SUPERMERCADOS[2], "Mundial");
    strcpy(SUPERMERCADOS[3], "Prezunic");
    strcpy(SUPERMERCADOS[4], "Zona Sul");

    //  PRECOS
    //  Carrefour
    PRECOS[0][0] = 5.10;
    PRECOS[0][1] = 4.98;
    PRECOS[0][2] = 5.00;
    PRECOS[0][3] = 5.80;
    PRECOS[0][4] = 23.12;
    PRECOS[0][5] = 3.99;
    PRECOS[0][6] = 12.00;
    PRECOS[0][7] = 5.87;
    PRECOS[0][8] = 5.90;
    PRECOS[0][9] = 5.60;

    //  Guanabara
    PRECOS[1][0] = 4.98;
    PRECOS[1][1] = 4.89;
    PRECOS[1][2] = 5.10;
    PRECOS[1][3] = 6.22;
    PRECOS[1][4] = 23.10;
    PRECOS[1][5] = 3.87;
    PRECOS[1][6] = 12.10;
    PRECOS[1][7] = 5.98;
    PRECOS[1][8] = 6.10;
    PRECOS[1][9] = 5.80;

    // Mundial
    PRECOS[2][0] = 5.02;
    PRECOS[2][1] = 5.10;
    PRECOS[2][2] = 4.98;
    PRECOS[2][3] = 6.10;
    PRECOS[2][4] = 23.04;
    PRECOS[2][5] = 4.10;
    PRECOS[2][6] = 12.05;
    PRECOS[2][7] = 6.00;
    PRECOS[2][8] = 5.98;
    PRECOS[2][9] = 5.98;

    //  Prezunic
    PRECOS[3][0] = 4.87;
    PRECOS[3][1] = 4.90;
    PRECOS[3][2] = 4.80;
    PRECOS[3][3] = 5.98;
    PRECOS[3][4] = 23.16;
    PRECOS[3][5] = 4.05;
    PRECOS[3][6] = 12.18;
    PRECOS[3][7] = 6.10;
    PRECOS[3][8] = 5.97;
    PRECOS[3][9] = 4.98;

    //  Zona sul
    PRECOS[4][0] = 5.15;
    PRECOS[4][1] = 5.15;
    PRECOS[4][2] = 4.99;
    PRECOS[4][3] = 6.20;
    PRECOS[4][4] = 21.96;
    PRECOS[4][5] = 4.38;
    PRECOS[4][6] = 11.97;
    PRECOS[4][7] = 6.15;
    PRECOS[4][8] = 5.96;
    PRECOS[4][9] = 5.30;

    return 0;
}

//  Esta função exibe o valor da variavel global CESTA_BASICA
int exibirCestaBasica(void) {
    for (int i = 0; i < 10; i++)
    {
        //  O comando %s é usado em conjunto com a biblioteca string.h para exibir strings no console
        printf("[%i] %s \n", i, CESTA_BASICA[i]);
    }
    return 0;
}

//  Esta função exibe o valor da variavel global SUPERMERCADO
int exibirSupermercados(void) {
    for (int i = 0; i < 5; i++)
    {
        printf("[%i] %s \n", i, SUPERMERCADOS[i]);
    }
    return 0;
}

//  Esta função exibe o valor da variavel global PRECOS
int exibirPrecos(void) {
    for (int linha = 0; linha < 5; linha++)
    {
        for (int coluna = 0; coluna < 10; coluna++)
        {
            printf("[%i][%i] %.2f \n", linha, coluna, PRECOS[linha][coluna]);
        }
    }
    return 0;
}

//  Esta função calcula o preço médio de cada produto
int calcularPrecoMedio(void) {
    double preco_medio[10];
    double soma = 0;

    for (int coluna = 0; coluna < 10; coluna++)
    {
        for (int linha = 0; linha < 5; linha++)
        {
            // Soma o preço de um produto em todos os supermercados
            soma = soma + PRECOS[linha][coluna];
        }
        // Após percorrer cada supermercado somando o preço do produto,
        // o valor somado é dividido pelo total de supermercados
        preco_medio[coluna] = soma/5;
        printf("%s -> Preco medio: R$%.2f \n", CESTA_BASICA[coluna], preco_medio[coluna]);
        soma = 0;
    }
    
}

//  Esta função calcula o total de cada cesta básica
int calcularCestaBasica(void) {
    double soma = 0;

    //  Este for percorre cada supermercado
    for (int linha = 0; linha < 5; linha++)
    {
        // Dentro de cada supermercado este for soma o valor total dos produtos que compõem a cesta básica
        for (int coluna = 0; coluna < 10; coluna++)
        {
            soma = soma + PRECOS[linha][coluna];
        }

        //  Aqui é exibido o nome do supermercado e o total de sua cesta básica
        printf("[%d] %s -> R$%.2f \n", linha, SUPERMERCADOS[linha], soma);

        //  A variável soma é zerada para ser usada nos cálculos do próximo supermercado.
        soma = 0;
    }

    return 0;
}

// Esta função indica qual supermercado oferece e a cesta mais barata e seu preço
int cestaMaisBarata(void) {
    double soma = 0, mais_barata = 1000;
    char *supermercado_mais_barato = "?";

    //  Este for percorre todos os supermercados
    for (int linha = 0; linha < 5; linha++)
    {
        //  Em cada supermercado soma o total de seus produtos
        for (int coluna = 0; coluna < 10; coluna++)
        {
            soma = soma + PRECOS[linha][coluna];
        }
        //  Se o total de produtos deste mercado for menor do que o valor armazenado
        //  na variável mais_barata então esta variável é atualizada, recebendo o novo menor valor
        //  da mesma forma, que o nome do supermercado mais barato é atualizado,
        //  passando como referência o valor contido no vetor SUPERMERCADO que seja compatível com a linha da matriz
        //  de preços onde achamos este novo valor mais barato
        if (soma < mais_barata)
        {
            mais_barata = soma;
            supermercado_mais_barato = SUPERMERCADOS[linha];
        }
        soma = 0;
    }
    
    printf("A cesta basica mais barata custa R$%.2f e eh vendida pelo supermercado %s\n", mais_barata, supermercado_mais_barato);
    return 0;
}

// Indicar o maior e menor preço de cada produto e onde ocorre
int precosMaxEMin(void) {
    //Variáveis de controle da função.
    double mais_caro = 1, mais_barato = 99;
    char *onde_e_caro = "", *onde_e_barato = "";

    //  Este loop percorre toda a coluna de produtos
    for (int coluna = 0; coluna < 10; coluna++)
    {
        //  Este loop percorre todos os preços de cada produto.
        for (int linha = 0; linha < 5; linha++)
        {
            //  Se o valor contido na posição do vetor PREÇOS referente as coordenadas atuais
            //  for maior do que o atual preço mais caro, este é atualizado recebendo o novo valor
            //  da mesma forma que o nome do supermercado onde este preço se encontra é atualizado.
            if (PRECOS[linha][coluna] >= mais_caro)
            {
                mais_caro = PRECOS[linha][coluna];
                onde_e_caro = SUPERMERCADOS[linha];
            }
            
            //  Por outro lado, se o valor contido na posição do vetor PREÇOS referente as coordenadas atuais
            //  for menor do que o atual preço mais barato, este é atualizado recebendo o novo valor
            //  da mesma forma que o nome do supermercado onde este preço se encontra é atualizado.
            if (PRECOS[linha][coluna] <= mais_barato)
            {
                mais_barato = PRECOS[linha][coluna];
                onde_e_barato = SUPERMERCADOS[linha];
            }
            
        }
        //  Após percorrer todas as variações de preço para o mesmo produto, as suas estatísticas são exibidas na tela
        //  E os valores das variáveis de controle da função são atualiados para os valores padrão.
        printf("Variacao do %s: de R$%.2f (%s) -> R$%.2f (%s) \n", CESTA_BASICA[coluna], mais_barato, onde_e_barato, mais_caro, onde_e_caro);
        mais_caro = 1;
        mais_barato = 99;
        onde_e_barato = "";
        onde_e_caro = "";
    }
    return 0;
}


// A função main apenas executa as funções do programa
int main(void) {

    //  Nunca comente a função inicializar, 
    //  ela precisa ser executada para que todas as demais funcionem corretamente.
    inicializar();

    //  Você pode comentar e descomentar as funções abaixo livremente,
    //  pois elas funcionam de forma independente entre si (mas todas dependem da função inicializar).
    //  Pode testá-las individualmente ou várias ao mesmo tempo.
    // exibirCestaBasica();
    // exibirSupermercados();
    // exibirPrecos();
    // calcularPrecoMedio();
    // calcularCestaBasica();
    // cestaMaisBarata();
    // precosMaxEMin();
    return 0;
}

File: 2-Semestre/AlgoritmosII/APS/aps.js
//MEMBROS DO GRUPO
//  Joao Pedro Hammes Pacheco (2024100014)
//  Membro 2
//  Membro 3
//  Membro 4

//• Armazenar os nomes dos 10 produtos da cesta básica que serão pesquisados
//      Cada ítem está armazenado em um índice da cesta básica, este índice serve como código do produto.
const Package = ['Arroz', 'Feijão', 'Açúcar', 'Óleo', 'Café', 'Fubá', 'Ovos', 'Macarrão', 'Leite', 'Frango']

//      Cada supermercado recebe um identificador
const Supermarkets = ['Carrefour', 'Guanabara', 'Mundial', 'Prezunic', 'Zona Sul']

//      Os produtos (referenciados pelos respectivos índices) possuem um valor em cada supermercado.
const carrefour = [5.10, 4.98, 5.00, 5.80, 23.12, 3.99, 12.00, 5.87, 5.90, 5.60]
const guanabara = [4.98, 4.89, 5.10, 6.22, 23.10, 3.87, 12.10, 5.98, 6.10, 5.80]
const mundial = [5.02, 5.10, 4.98, 6.10, 23.04, 4.10, 12.05, 6.00, 5.98, 5.98]
const prezunic = [4.87, 4.90, 4.80, 5.98, 23.16, 4.05, 12.18, 6.10, 5.97, 4.98]
const zona_sul = [5.15, 5.15, 4.99, 6.20, 21.96, 4.38, 11.97, 6.15, 5.96, 5.30]

//• Armazenar os nomes dos 5 supermercados onde será realizada a pesquisa de preço.
//      Todos os supermercados foram armazenados em um mesmo vetor,
//      de forma que seja formado um vetor de vetores, ou seja: uma matriz.
const PackagesBySupermarkets = [
    carrefour,
    guanabara,
    mundial,
    prezunic,
    zona_sul
]

//• Ler os preços dos 10 produtos pesquisados em cada supermercado.

// • Calcular o preço médio de cada produto (10) da cesta básica nos 5 supermercados pesquisados.
function calculateAverageProductPrice() {

    //  Inicializa o preço médio de cada produto em 0.
    let AveragePrice = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

    //  Estes loops aninhados são responsáveis por somar o preço de cada produto (coluna)
    //  em cada supermercado (linha).
    //      Este loop externo percorre cada produto (coluna) armazenado em Package.
    for (let column = 0; column < Package.length; column++) {
        console.log(`${Package[column]}`)

        //  Este loop interno SOMA o preço deste produto em cada supermercado (linha)
        for (let line = 0; line < PackagesBySupermarkets.length; line++) {
            AveragePrice[column] += PackagesBySupermarkets[line][column]
        }
        
    //  Assim que o algoritmo termina de somar o preço do produto em todos os supermercados (linhas)
    //  ele calcula a média, dividindo o total pela quantidade de supermercados (linhas).
        AveragePrice[column] = Number(AveragePrice[column]/PackagesBySupermarkets.length).toFixed(2)
        console.log(`Preço médio: R$${AveragePrice[column]}\n`)
    }
}

// • Calcular o valor da cesta básica nos 5 supermercados pesquisados (Soma).
function calculatePackageByMarket() {
    
    //  Inicializo o preço total de cada cesta básica em 0
    let PackageByMarket = [0, 0, 0, 0, 0]

    //  O loop abaixo percorre cada supermercado (linha) armazenado no vetor PackagesBySupermarkets
    for (let supermarket = 0; supermarket < PackagesBySupermarkets.length; supermarket++) {

        //  O loop abaixo percorre cada produto (coluna) do supermercado selecionado no loop acima
        for (let product = 0; product < PackagesBySupermarkets[supermarket].length; product++) {

            //  A linha abaixo é responsável somar o valor da cesta básica do supermercado
            //  incrementando o valor já somado com o valor do produto atual
            PackageByMarket[supermarket] += PackagesBySupermarkets[supermarket][product]
        }
        //  A linha abaixo garante que o valor somado na cesta básica do supermercado seja tipo numérico
        //  e tenha 2 valores após a vírgula para representar adequadamente a moeda local.
        PackageByMarket[supermarket] = Number(PackageByMarket[supermarket]).toFixed(2)
    }

    //  Ao final do loops é retornado no console o valor da cesta básica em cada supermercado
    //  Usando como referência o nome do supermercado alocado em supermarkets
    //  e passando como Preço os valores armazenados em PackageByMarket
    console.log('Preço da cesta básica em cada supermercado:')
    for (let column = 0; column < PackageByMarket.length; column++) {
        console.log(`${Supermarkets[column]}: ${PackageByMarket[column]}`)        
    }    
}

// • Indicar qual o menor valor da cesta básica e em qual supermercado ele ocorre.
function lessExpensivePackage() {
    
    //  Inicializa o preço total de cada cesta básica em 0
    let PackagePrice = [0, 0, 0, 0, 0]

    //  Percorre cada supermercado (linha) e soma o total de seus produtos (coluna)
    //  Armazena a soma total em PackagePrice em formato BRL.
    for (let supermarket = 0; supermarket < PackagesBySupermarkets.length; supermarket++) {
        for (let product = 0; product < PackagesBySupermarkets[supermarket].length; product++) {
            PackagePrice[supermarket] += PackagesBySupermarkets[supermarket][product]
        }
        PackagePrice[supermarket] = Number(PackagePrice[supermarket]).toFixed(2)
    }

    //  Define valores padrão iniciais para as variáveis que definem
    //  o preço mais baixo e em que supermercado ele ocorre
    let lowerPrice = 999.99
    let where = '?'

    //  Percorre o array de preços das cestas básicas
    for (let column = 0; column < PackagePrice.length; column++) {
        //  Se a cesta básica atual tive um preço menor do que o valor armazenado
        //  em lowerPrice, então o valor de lowerPrice é atualizado
        //  E o valor do supermercado onde este preço está é atualizado também,
        //  passando como valor o supermercado armazenado na posição column do vetor de supermercados
        if (PackagePrice[column] < lowerPrice) {
            lowerPrice = PackagePrice[column]
            where = Supermarkets[column]
        }
    }

    return console.log(`O preço mais baixo encontrado para cesta básica foi:\n ${where} - ${lowerPrice}`)
}

// • Determinar para cada produto: o preço mínimo e o preço máximo
// e indicar em qual supermercado ele ocorre.
function higherLowerWhere() {
    //  Define valor padrão para as seguintes características cada produto:
    //  mais barato e onde está, mais caro e onde está
    let more_expensive = 0.01
    let cheaper =  999.99
    let where_is_more_expensive = ''
    let where_is_cheaper = ''

    //  Percorre cada produto (coluna)
    for (let product = 0; product < Package.length; product++) {
        
        //  E confere seu preço em cada supermercado (linha)
        for (let supermarket = 0; supermarket < PackagesBySupermarkets.length; supermarket++) {
            
            //  Se o valor nesta posição da matriz for maior do que o contido na variável more_expensive substitua,
            //  atualizando o valor do maior preço deste produto e formatando-o para a um número com duas casas decimais.
            //  Em seguida atualiza também o valor do supermercado onde este preço ocorre.
            if (PackagesBySupermarkets[supermarket][product] >= more_expensive) {
                more_expensive = Number(PackagesBySupermarkets[supermarket][product]).toFixed(2)
                where_is_more_expensive = Supermarkets[supermarket]
            }

            //  Por outro lado, se o valor nesta posição da matriz for menor do que o contido na variável cheaper substitua,
            //  atualizando o valor do menor preço deste produto e formatando-o para a um número com duas casas decimais.
            //  Em seguida atualiza também o valor do supermercado onde este preço ocorre.
            if (PackagesBySupermarkets[supermarket][product] <= cheaper) {
                cheaper = Number(PackagesBySupermarkets[supermarket][product]).toFixed(2)
                where_is_cheaper = Supermarkets[supermarket]
            }
        }

        //  Exibe o maior e o menor preço de cada produto
        console.log(`\n${Package[product]}\nMenor preço: R$${cheaper} (${where_is_cheaper}) | Maior preço: R$${more_expensive} (${where_is_more_expensive})  \n`)

        //  Atualiza os valores padrão de bigger e smaller para serem usados no próximo produto.
        more_expensive = 0.01
        cheaper =  999.99
        where_is_more_expensive = ''
        where_is_cheaper = ''
    }
}

//  AREA DE TESTES
//calculateAverageProductPrice()
//calculatePackageByMarket()
//lessExpensivePackage()
higherLowerWhere()